### PR TITLE
Feature/59-city-crud

### PR DESCRIPTION
**Pull request by adCBueno**

### CODE REVIEW

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/view/CityAdminDetailFragment.kt [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                val countyId = viewModel.cityCountry?.value?.get(binding.spCategory.selectedItemPosition)?.id
+                viewModel.saveNewCity(name, lat, lng, selectedCountry, countyId ?: "0")
+            } else {
+                showProgress(true)
+                val name = binding.etName.text.toString()
+                val lat = binding.etLat.text.toString()
+                val lng = binding.etLng.text.toString()
+                val selectedCountry = (binding.spCategory.selectedItem as Country).name
+                val countyId = viewModel.cityCountry?.value?.get(binding.spCategory.selectedItemPosition)?.id
+                viewModel.updateCity(name, lat, lng, selectedCountry, countyId ?: "0", city!!)
```
Comment (Rodrigo-Schar) [inline app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/view/CityAdminDetailFragment.kt]: don't force to wrap the city, validate it firts

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        var isValid = ""
+        if (name.isEmpty()) {
+            isValid = RoutesAppApplication.resource?.getString(R.string.reg_val_name).toString()
+            return isValid
+        }
+        return isValid
+    }
+
+    fun saveNewCity(name: String, latitude: String, longitude: String, countrySelected: String, countryId: String) = viewModelScope.launch {
+        val valid = validateFields(name)
```
Comment (Rodrigo-Schar) [inline app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt]: you have more fields necesary like lat and lng so make sure to validate them too

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+class CityAdminViewModel
+@Inject
+constructor(private val cityRepository: CityRepository) : ViewModel() {
+    var _cityList: MutableLiveData<List<City>> = MutableLiveData()
+    val cityList: LiveData<List<City>> = _cityList
+    var originalList: List<City> = listOf()
+    var searchQuery = ""
+
+    private var _cityCountry: MutableLiveData<List<Country>> = MutableLiveData()
+    val cityCountry: LiveData<List<Country>> = _cityCountry // val lineCategories: LiveData<List<LineCategories>> = _lineCategories
```
Comment (Rodrigo-Schar) [inline app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt]: delete unnecesary comments

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/view/CityAdminDetailFragment.kt (hunk 1 above)
Comment (adCBueno) [reply app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/view/CityAdminDetailFragment.kt]: Corrected :)

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt (hunk 2 above)
Comment (adCBueno) [reply app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt]: Corrected :)

File: app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt (hunk 3 above)
Comment (adCBueno) [reply app/src/main/java/com/jalasoft/routesapp/ui/adminPages/cities/viewModel/CityAdminViewModel.kt]: Corrected :)

**Review by Rodrigo-Schar — APPROVED** [review]

**Review by jmacboy — APPROVED** [review]

**Pull request merged**